app/api/items/ask_hint.feature

```diff
@@ -0,0 +1,227 @@
+Feature: Ask for a hint
+  Background:
+    Given the database has the following table 'users':
+      | ID  | sLogin | idGroupSelf |
+      | 10  | john   | 101         |
+    And the database has the following table 'groups':
+      | ID  |
+      | 101 |
+    And the database has the following table 'groups_ancestors':
+      | idGroupAncestor | idGroupChild | bIsSelf |
+      | 101             | 101          | 1       |
+    And the database has the following table 'groups_groups':
+      | ID | idGroupParent | idGroupChild | sType              | sStatusDate |
+      | 15 | 22            | 13           | direct             | null        |
+    And the database has the following table 'platforms':
+      | ID | bUsesTokens | sRegexp                                           | sPublicKey                |
+      | 10 | 1           | http://taskplatform.mblockelet.info/task.html\?.* | {{taskPlatformPublicKey}} |
+    And the database has the following table 'items':
+      | ID | idPlatform | sUrl                                                                    |
+      | 50 | 10         | http://taskplatform.mblockelet.info/task.html?taskId=403449543672183936 |
+      | 10 | null       | null                                                                    |
+    And the database has the following table 'items_items':
+      | idItemParent | idItemChild |
+      | 10           | 50          |
+    And the database has the following table 'items_ancestors':
+      | idItemAncestor | idItemChild |
+      | 10             | 50          |
+    And the database has the following table 'users_items':
+      | idUser | idItem | sHintsRequested                 | nbHintsCached | nbSubmissionsAttempts | idAttemptActive |
+      | 10     | 50     | [{"rotorIndex":0,"cellRank":0}] | 12            | 2                     | 100             |
+      | 10     | 10     | null                            | 0             | 0                     | null            |
+    And time is frozen
+
+  Scenario: User is able to ask for a hint
+    Given I am the user with ID "10"
+    When I send a POST request to "/items/ask_hint" with the following body:
+      """
+      {
+        "task_token": {{generateToken(map(
+          "idUser", "10",
+          "idItemLocal", "50",
+          "platformName", app().TokenConfig.PlatformName,
+        ), app().TokenConfig.PrivateKey)}},
```

Comment:
Token dates handling is completely covered by the token.TestParseAndValidate() test. There is no need to test it again with bdd.

Reply:
> `And the "task_token" from the response data is a the following token that can be validated with "app().TokenConfig.PublicKey"`
The response data is not a token, but a JSON with a token inside.

Reply:
> > `And the "task_token" from the response data is a the following token that can be validated with "app().TokenConfig.PublicKey"`
> 
> The response data is not a token, but a JSON with a token inside.

That is why there 1 check that this is a valid created response and then that the value at the "task_token" key is the expected token. Without the need for function call inside the output format which does not ensure the reader about the actual content.

Reply:
Actually, in case you would not have understood already, my remark is not really about the correctness of your code or test coverage (I am sure it is excellent). It is about reading the bdd tests as an external person (not the one who has implemented this) and being convinced that it is what was expected without scratching one's head for a few minutes or reading the test helpers code.